ably/realtime_conn_spec_test.go

```diff
@@ -3562,16 +3518,439 @@ func TestRealtimeConn_RTC8a_ExplicitAuthorizeWhileConnected(t *testing.T) {
+		auth := newTestAuthCallbackHandler()
+		// We'll use this REST to get real, working tokens.
+		app, rest := ablytest.NewREST()
+		defer safeclose(t, app)
+
+		var expectedToken, gotToken string
+
+		// Set a DialFunc that lets us make connection attempts fail.
+		gotDial := make(chan chan error, 1)
+		var lastDialURL *url.URL
+		dial := func(protocol string, u *url.URL, timeout time.Duration) (ably.Conn, error) {
+			dialErr := make(chan error)
+			lastDialURL = u
+			gotDial <- dialErr
+			err := <-dialErr
+			if err != nil {
+				return nil, err
+			}
+			return ably.DialWebsocket(protocol, u, timeout)
+		}
+		dial, intercept := DialIntercept(dial)
+		c := app.NewRealtime(
+			ably.WithDial(dial),
+			ably.WithAutoConnect(false),
+			ably.WithAuthCallback(auth.authCallback),
+		)
+		defer safeclose(t, ablytest.FullRealtimeCloser(c))
+
+		// Move to CONNECTING; call Authorize while there.
+		c.Connect()
+		var finishDial chan error
+		auth.handleAuth(t, getRealToken(t, rest), nil)
+		ablytest.Instantly.Recv(t, &finishDial, gotDial, t.Fatalf)
+
+		// Now call Authorize.
+		authorizeDone := make(chan struct{})
+		var rg ablytest.ResultGroup
+		defer rg.Wait()
+		rg.GoAdd(func(ctx context.Context) error {
+			defer close(authorizeDone)
+			_, err := c.Auth.Authorize(ctx, nil)
+			return err
+		})
+		auth.handleAuth(t, func(params ably.TokenParams) ably.Tokener {
+			t := getRealToken(t, rest)(params)
+			expectedToken = t.(*ably.TokenDetails).Token
+			return t
+		}, nil)
+		// Wait a bit to ensure we don't move to DISCONNECTED before
+		// onExplicitAuthorize sees that the current state is CONNECTING.
+		// (Doing this deterministically would need more hooks inside
+		// onExplicitAuthorize. Can be done, but probably not worth the trouble.)
+		time.Sleep(100 * time.Millisecond)
+
+		// Move to DISCONNECTED. This should then try to connect again right
+		// away, which will use the new token.
+		finishDial <- errors.New("test")
+
+		// Intercept CONNECTED, to check that Authorize doesn't return
+		// before the reauthorization it initiated finishes (RTC8b1).
+		ablytest.Instantly.NoRecv(t, nil, authorizeDone, t.Fatalf)
+		ctx, stopIntercepting := context.WithCancel(context.Background())
+		defer stopIntercepting()
+		connectedMsg := intercept(ctx, ably.ActionConnected)
+
+		// Expect a new dial. Check that the URL carries the new token.
+		ablytest.Soon.Recv(t, &finishDial, gotDial, t.Fatalf)
+		gotToken = lastDialURL.Query().Get("access_token")
+		finishDial <- nil
+
+		ablytest.Instantly.NoRecv(t, nil, authorizeDone, t.Fatalf)
+
+		// Now let the reauthorization finish.
+		stopIntercepting()
+		ablytest.Soon.Recv(t, nil, connectedMsg, t.Fatalf)
+		ablytest.Instantly.Recv(t, nil, authorizeDone, t.Fatalf)
+
+		assertEquals(t, expectedToken, gotToken)
+	})
+}
+
+func TestRealtimeConn_RTC8c_ExplicitAuthorizeConnects(t *testing.T) {
+	t.Parallel()
+
+	test := func(t *testing.T, initialState ably.ConnectionState) {
```

Comment:
Can we rename `test` to `authorizeWithConnectionState`, and then pass the right state to the function